### ISSUE TITLE
Missing syntax highlighting for multi variable declaration

### ISSUE THREAD

**Issue by abbasyadollahi** [microsoft/pylance-release#3904]
Type: <b>Bug</b>

# Behaviour
### Actual

![image](https://user-images.githubusercontent.com/16931654/216478510-0b18540c-bc91-4ed5-8173-49b8a42a677e.png)

In the first case, only the `right` variable is highlighted.
In the second case, none of the variables are highlighted.

### Expected

The same content as above but with all 3 `left`, `middle` and `right` variables highlighted.

## Steps to reproduce:

1. Set multiple variables in a single statement over a single or multiple lines.

# Diagnostic data

-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.6
-   Type of virtual environment used (e.g. conda, venv, virtualenv, etc.): Global
-   Value of the `python.languageServer` setting: Pylance

<details>

<summary>Output for <code>Python</code> in the <code>Output</code> panel (<code>View</code>→<code>Output</code>, change the drop-down the upper-right of the <code>Output</code> panel to <code>Python</code>)
</summary>

<p>

```
Telemetry is disabled
Telemetry level is off
Experiments are disabled, only manually opted experiments are active.
LSP Notebooks experiment is enabled
LSP Notebooks interactive window support is enabled
> conda info --json
Python interpreter path: /usr/local/bin/python3
> /usr/bin/python3 -I ~/.vscode/extensions/ms-python.python-2023.2.0/pythonFiles/get_output_via_markers.py ~/.vscode/extensions/ms-python.python-2023.2.0/pythonFiles/interpreterInfo.py
> /usr/local/bin/python3 -I ~/.vscode/extensions/ms-python.python-2023.2.0/pythonFiles/get_output_via_markers.py ~/.vscode/extensions/ms-python.python-2023.2.0/pythonFiles/interpreterInfo.py
Starting Pylance language server.
```

</p>
</details>

<details>

<summary>User Settings</summary>

<p>

```

languageServer: "Pylance"

```

</p>
</details>

Extension version: 2023.2.0
VS Code version: Code 1.75.0 (Universal) (e2816fe719a4026ffa1ee0189dc89bdfdbafb164, 2023-02-01T15:24:42.903Z)
OS version: Darwin x64 22.3.0
Modes:
Sandboxed: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz (12 x 2600)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|5, 3, 3|
|Memory (System)|16.00GB (0.01GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|
</details>

**Comment by debonte** [comment]
In these scenarios we're sending negative delta line (`pushLine`) and delta char (`pushChar`) values back in our semantic tokens response. This is happening because the parse tree walk visits the names in each multi-variable assignment from right to left.

Here are the arguments from the relevant calls to `SemanticTokensBuilder.push`, in the order that they occur, plus the resulting `pushLine` and `pushChar` values:
|name|`line`|`char`|`length`|`tokenType`|`tokenModifiers`|`pushLine`|`pushChar`|
|:----:|:----:|:----:|:------:|:---------:|:--------------:|:--------:|:--------:|
|right|0|16|5|variable|declaration|0|16|
|middle|0|7|6|variable|declaration|0|-9|
|left|0|0|4|variable|declaration|0|-7|
|right|4|8|5|variable|none|4|8|
|middle|3|4|6|variable|none|-1|4|
|left|2|0|4|variable|none|-1|0|

If the order is reversed so that all of the `pushLine` and `pushChar` values are non-negative, semantic highlighting works as expected.

I noticed that `tokenModifiers` on the multi-line declarations is set to `none` instead of `declaration` which I believe is also incorrect. Might want to fix that at the same time.

**Comment by erictraut** [comment]
This might be fallout from a bug fix I introduced in late December: https://github.com/microsoft/pyright/issues/4368. In most languages (like C, Java, etc.), chained assignments are evaluated right to left. For example, if you have `a = b = c = 3`, the assignment to `c` occurs before the assignment to `b`, etc. It turns out that in Python, it's the other way around. Normally, the order doesn't matter, but it can in some cases, like `a = z[a] = 0`.

My fix was to reformulate the parse tree to reflect the assignment ordering. We may need to rethink this. Or change the semantic token logic to special case chained assignment expressions.

**Comment by debonte** [comment]
Sounds likely. This regressed in Pylance [2023.1.11](https://github.com/microsoft/pyrx/blob/main/packages/vscode-pylance/CHANGELOG.md#2023111-4-january-2023-prerelease), which upgraded from Pyright 1.1.283 to 1.1.287, so it brought in a lot of Pyright changes including that one.

**Comment by bschnurr** [comment]
This issue has been fixed in prerelease version 2023.2.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023241-22-February-2023-PreRelease)
